# Topic 11b — KBinsDiscretizer (Discretization) — ✍️ Practice & Solutions

This notebook has every **practice problem** for Topic 11b, each followed by its **✅ solution**.

**How to use it**
1. Read the lesson first: [Topic_11b_Lesson.ipynb](Topic_11b_Lesson.ipynb).
2. Read a problem and write your answer in the code cell under it.
3. Only then scroll down to the ✅ solution and compare.
4. Stuck or confused? Go back to the matching lesson section — every solution uses only what the lesson teaches.

> 📚 **Course navigation** — you are in **Topic 11b: KBinsDiscretizer (Discretization)**
> ⬅️ Previous: [Topic 11 — Feature Engineering](../Topic_11_Feature_Engineering/Topic_11_Lesson.ipynb)  ·  ➡️ Next: [Topic 12 — Correlation & Multicollinearity](../Topic_12_Correlation_Multicollinearity/Topic_12_Lesson.ipynb)  ·  🗂️ [All topics](../Topic_00_All_Topics_Index.ipynb)

<details><summary>Full course sequence (click to expand)</summary>

| # | Topic | Lesson | Practice & Solutions |
|---|---|---|---|
| 01 | Python Basics | [📖 Lesson](../Topic_01_Python_Basics/Topic_01_Lesson.ipynb) | [✍️ Practice](../Topic_01_Python_Basics/Topic_01_Practice_Solutions.ipynb) |
| 02 | Pandas DataFrame Basics | [📖 Lesson](../Topic_02_Pandas_DataFrame_Basics/Topic_02_Lesson.ipynb) | [✍️ Practice](../Topic_02_Pandas_DataFrame_Basics/Topic_02_Practice_Solutions.ipynb) |
| 03 | Handling Missing Values | [📖 Lesson](../Topic_03_Handling_Missing_Values/Topic_03_Lesson.ipynb) | [✍️ Practice](../Topic_03_Handling_Missing_Values/Topic_03_Practice_Solutions.ipynb) |
| 04 | Normalization & Standardization | [📖 Lesson](../Topic_04_Normalization_Standardization/Topic_04_Lesson.ipynb) | [✍️ Practice](../Topic_04_Normalization_Standardization/Topic_04_Practice_Solutions.ipynb) |
| 05 | Detecting & Treating Outliers | [📖 Lesson](../Topic_05_Outliers/Topic_05_Lesson.ipynb) | [✍️ Practice](../Topic_05_Outliers/Topic_05_Practice_Solutions.ipynb) |
| 06 | Encoding Techniques | [📖 Lesson](../Topic_06_Encoding_Techniques/Topic_06_Lesson.ipynb) | [✍️ Practice](../Topic_06_Encoding_Techniques/Topic_06_Practice_Solutions.ipynb) |
| 07 | Exploratory Data Analysis (EDA) | [📖 Lesson](../Topic_07_Exploratory_Data_Analysis/Topic_07_Lesson.ipynb) | [✍️ Practice](../Topic_07_Exploratory_Data_Analysis/Topic_07_Practice_Solutions.ipynb) |
| 08 | Data Visualization | [📖 Lesson](../Topic_08_Data_Visualization/Topic_08_Lesson.ipynb) | [✍️ Practice](../Topic_08_Data_Visualization/Topic_08_Practice_Solutions.ipynb) |
| 09 | Pandas Data Manipulation | [📖 Lesson](../Topic_09_Pandas_Data_Manipulation/Topic_09_Lesson.ipynb) | [✍️ Practice](../Topic_09_Pandas_Data_Manipulation/Topic_09_Practice_Solutions.ipynb) |
| 10 | Combining DataFrames | [📖 Lesson](../Topic_10_Combining_DataFrames/Topic_10_Lesson.ipynb) | [✍️ Practice](../Topic_10_Combining_DataFrames/Topic_10_Practice_Solutions.ipynb) |
| 11 | Feature Engineering | [📖 Lesson](../Topic_11_Feature_Engineering/Topic_11_Lesson.ipynb) | [✍️ Practice](../Topic_11_Feature_Engineering/Topic_11_Practice_Solutions.ipynb) |
| **▶ 11b ◀** | **KBinsDiscretizer (Discretization)** | [📖 Lesson](../Topic_11b_KBinsDiscretizer/Topic_11b_Lesson.ipynb) | [✍️ Practice](../Topic_11b_KBinsDiscretizer/Topic_11b_Practice_Solutions.ipynb) |
| 12 | Correlation & Multicollinearity | [📖 Lesson](../Topic_12_Correlation_Multicollinearity/Topic_12_Lesson.ipynb) | [✍️ Practice](../Topic_12_Correlation_Multicollinearity/Topic_12_Practice_Solutions.ipynb) |
| 13 | Data Transformation & Skewness | [📖 Lesson](../Topic_13_Data_Transformation_Skewness/Topic_13_Lesson.ipynb) | [✍️ Practice](../Topic_13_Data_Transformation_Skewness/Topic_13_Practice_Solutions.ipynb) |
| 14 | Statistics for Data Science | [📖 Lesson](../Topic_14_Statistics_for_Data_Science/Topic_14_Lesson.ipynb) | [✍️ Practice](../Topic_14_Statistics_for_Data_Science/Topic_14_Practice_Solutions.ipynb) |
| 15 | Probability for Data Science & ML | [📖 Lesson](../Topic_15_Probability_for_Data_Science_ML/Topic_15_Lesson.ipynb) | [✍️ Practice](../Topic_15_Probability_for_Data_Science_ML/Topic_15_Practice_Solutions.ipynb) |
| 16 | Feature Selection | [📖 Lesson](../Topic_16_Feature_Selection/Topic_16_Lesson.ipynb) | [✍️ Practice](../Topic_16_Feature_Selection/Topic_16_Practice_Solutions.ipynb) |
| 17 | Machine Learning | [📖 Lesson](../Topic_17_Machine_Learning/Topic_17_Lesson.ipynb) | [✍️ Practice](../Topic_17_Machine_Learning/Topic_17_Practice_Solutions.ipynb) |

</details>

## 📋 Practice Index

**30 problems** — each one is followed by your answer cell and then its ✅ solution. Click a problem to jump to it.

- [✍️ Practice Problem 1](#p-1)
- [✍️ Practice Problem 2](#p-2)
- [✍️ Practice Problem 3](#p-3)
- [✍️ Practice Problem 4](#p-4)
- [✍️ Practice Problem 5](#p-5)
- [✍️ Practice Problem 6](#p-6)
- [✍️ Practice Problem 7](#p-7)
- [✍️ Practice Problem 8](#p-8)
- [✍️ Practice Problem 9](#p-9)
- [✍️ Practice Problem 10](#p-10)
- [✍️ Practice Problem 11](#p-11)
- [✍️ Practice Problem 12](#p-12)
- [✍️ Practice Problem 13](#p-13)
- [✍️ Practice Problem 14](#p-14)
- [✍️ Practice Problem 15](#p-15)
- [✍️ Practice Problem 16](#p-16)
- [✍️ Practice Problem 17](#p-17)
- [✍️ Practice Problem 18](#p-18)
- [✍️ Practice Problem 19](#p-19)
- [✍️ Practice Problem 20](#p-20)
- [✍️ Practice Problem 21](#p-21)
- [✍️ Practice Problem 22](#p-22)
- [✍️ Practice Problem 23](#p-23)
- [✍️ Practice Problem 24](#p-24)
- [✍️ Practice Problem 25](#p-25)
- [✍️ Practice Problem 26](#p-26)
- [✍️ Practice Problem 27](#p-27)
- [✍️ Practice Problem 28](#p-28)
- [Section 31 — Interview Questions](#p-29)
- [Section 32 — Final Practice Assignment](#p-30)

> 📖 Lesson for this topic: [Topic_11b_Lesson.ipynb](Topic_11b_Lesson.ipynb)

### ⚙️ Setup — run this cell first

Some problems use data, variables and functions created in the lesson. This cell quietly runs the whole lesson notebook so they all exist here.

In [1]:
%%capture
# Runs the lesson notebook (output hidden) so its variables exist here
%run ./Topic_11b_Lesson.ipynb

<a id="p-1"></a>
### ✍️ Practice Problem 1

Create your own small DataFrame (6–10 rows) with one continuous column (for example `Temperature`, `Income`, or `Exam_Marks`). Use `pd.cut()` to create 3 meaningful groups with labels, then write 2–3 sentences in a comment: what information is lost, and where is a possible bin-boundary problem?

*(Solution in the **Practice Solutions** section at the end.)*

In [2]:
# ✍️ Practice Problem 1 — write your solution here

---
---
# Practice Solutions

⚠️ Try each problem yourself first! Each solution is followed by a short explanation.

### Solution 1

In [3]:
temps = pd.DataFrame({'Temperature': [-3, 2, 8, 12, 17, 21, 26, 31, 36, 40]})
temps['Feel'] = pd.cut(temps['Temperature'], bins=[-10, 10, 25, 50], labels=['Cold', 'Mild', 'Hot'])
temps
# Information lost: 12 °C and 21 °C are both just "Mild" (a 9-degree difference disappears).
# Boundary problem: 10 °C is "Cold" but 10.5 °C would be "Mild" — almost identical temperatures, different groups.

,Temperature,Feel
0,-3,Cold
1,2,Cold
2,8,Cold
3,12,Mild
4,17,Mild
5,21,Mild
6,26,Hot
7,31,Hot
8,36,Hot
9,40,Hot


**Explanation:** Any small dataset works. The key idea: grouping loses detail inside a bin, and values just either side of a boundary are treated very differently.

<a id="p-2"></a>
### ✍️ Practice Problem 2

Create a `KBinsDiscretizer` with **5 bins**, `encode='ordinal'` and `strategy='quantile'`. Print the object and its parameters with `get_params()`.

*(Solution in the **Practice Solutions** section at the end.)*

In [4]:
# ✍️ Practice Problem 2 — write your solution here

### Solution 2

In [5]:
kb2 = KBinsDiscretizer(n_bins=5, encode='ordinal', strategy='quantile')
print(kb2)
print(kb2.get_params())

KBinsDiscretizer(encode='ordinal')
{'dtype': None, 'encode': 'ordinal', 'n_bins': 5, 'quantile_method': 'averaged_inverted_cdf', 'random_state': None, 'strategy': 'quantile', 'subsample': 200000}


**Explanation:** Nothing is learned yet — the object only stores settings until `fit` is called.

<a id="p-3"></a>
### ✍️ Practice Problem 3

Create a DataFrame with a `Salary` column: `[25000, 32000, 40000, 45000, 52000, 61000, 75000, 90000, 120000, 150000]`. Discretize it into **4 bins** (`encode='ordinal'`, `strategy='uniform'`), add the bin IDs as a new column, and print `bin_edges_`.

*(Solution in the **Practice Solutions** section at the end.)*

In [6]:
# ✍️ Practice Problem 3 — write your solution here

### Solution 3

In [7]:
sal3 = pd.DataFrame({'Salary': [25000, 32000, 40000, 45000, 52000, 61000, 75000, 90000, 120000, 150000]})
kb3 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='uniform')
sal3['Salary_Bin'] = kb3.fit_transform(sal3[['Salary']]).astype(int).ravel()
print('Edges:', kb3.bin_edges_[0])
sal3

Edges: [ 25000.  56250.  87500. 118750. 150000.]


,Salary,Salary_Bin
0,25000,0
1,32000,0
2,40000,0
3,45000,0
4,52000,0
5,61000,1
6,75000,1
7,90000,2
8,120000,3
9,150000,3


**Explanation:** Range 25,000–150,000 → width 31,250. Five of ten people fall into bin 0 because salaries are right-skewed; the bin numbers are group IDs, not salaries.

<a id="p-4"></a>
### ✍️ Practice Problem 4

Create a DataFrame with a `Temperature` column (at least 10 values between −5 and 40 °C). Create **3 bins** with `strategy='uniform'`, then map the bin IDs to the labels `Cold`, `Mild`, `Hot`.

*(Solution in the **Practice Solutions** section at the end.)*

In [8]:
# ✍️ Practice Problem 4 — write your solution here

### Solution 4

In [9]:
temp4 = pd.DataFrame({'Temperature': [-5, -1, 3, 8, 12, 16, 20, 24, 29, 33, 37, 40]})
kb4 = KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='uniform')
temp4['Bin'] = kb4.fit_transform(temp4[['Temperature']]).astype(int).ravel()
temp4['Label'] = temp4['Bin'].map({0: 'Cold', 1: 'Mild', 2: 'Hot'})
print('Edges:', kb4.bin_edges_[0])
temp4

Edges: [-5. 10. 25. 40.]


,Temperature,Bin,Label
0,-5,0,Cold
1,-1,0,Cold
2,3,0,Cold
3,8,0,Cold
4,12,1,Mild
5,16,1,Mild
6,20,1,Mild
7,24,1,Mild
8,29,2,Hot
9,33,2,Hot


**Explanation:** Range −5 to 40 → 3 bins of width 15: [−5, 10), [10, 25), [25, 40].

<a id="p-5"></a>
### ✍️ Practice Problem 5

Create exam marks from 0 to 100 (e.g. `[0, 8, 15, 23, 37, 44, 52, 61, 68, 77, 85, 93, 100]`). Use `strategy='uniform'` with **5 bins**, print the edges and the widths, and map the bins to grades `E, D, C, B, A`.

*(Solution in the **Practice Solutions** section at the end.)*

In [10]:
# ✍️ Practice Problem 5 — write your solution here

### Solution 5

In [11]:
marks5 = pd.DataFrame({'Marks': [0, 8, 15, 23, 37, 44, 52, 61, 68, 77, 85, 93, 100]})
kb5 = KBinsDiscretizer(n_bins=5, encode='ordinal', strategy='uniform')
marks5['Bin'] = kb5.fit_transform(marks5[['Marks']]).astype(int).ravel()
marks5['Grade'] = marks5['Bin'].map({0: 'E', 1: 'D', 2: 'C', 3: 'B', 4: 'A'})
print('Edges :', kb5.bin_edges_[0]); print('Widths:', np.diff(kb5.bin_edges_[0]))
marks5

Edges : [  0.  20.  40.  60.  80. 100.]
Widths: [20. 20. 20. 20. 20.]


,Marks,Bin,Grade
0,0,0,E
1,8,0,E
2,15,0,E
3,23,1,D
4,37,1,D
5,44,2,C
6,52,2,C
7,61,3,B
8,68,3,B
9,77,3,B


**Explanation:** Because the data spans exactly 0–100, uniform bins are 20 marks wide — identical to a classic grading scale.

<a id="p-6"></a>
### ✍️ Practice Problem 6

Create a `Spending` column with skewed values, e.g. `[120, 150, 180, 200, 220, 260, 300, 350, 900, 2500, 8000, 15000]`. Use `strategy='quantile'` with **4 bins**. Print the edges and the number of customers in each group, then do the same with `uniform` and compare.

*(Solution in the **Practice Solutions** section at the end.)*

In [12]:
# ✍️ Practice Problem 6 — write your solution here

### Solution 6

In [13]:
spend6 = pd.DataFrame({'Spending': [120, 150, 180, 200, 220, 260, 300, 350, 900, 2500, 8000, 15000]})
for strat in ['quantile', 'uniform']:
    kb6 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy=strat)
    spend6[strat] = kb6.fit_transform(spend6[['Spending']]).astype(int).ravel()
    print(f'{strat:<9} edges {kb6.bin_edges_[0].round(0)} | customers per group {spend6[strat].value_counts().reindex(range(4), fill_value=0).tolist()}')
spend6

quantile  edges [  120.   190.   280.  1700. 15000.] | customers per group [3, 3, 3, 3]
uniform   edges [  120.  3840.  7560. 11280. 15000.] | customers per group [10, 0, 1, 1]


,Spending,quantile,uniform
0,120,0,0
1,150,0,0
2,180,0,0
3,200,1,0
4,220,1,0
5,260,1,0
6,300,2,0
7,350,2,0
8,900,2,0
9,2500,3,0


**Explanation:** Quantile gives 3 customers per group; uniform puts 10 of 12 customers in group 0 because two big spenders stretch the range.

<a id="p-7"></a>
### ✍️ Practice Problem 7

Create a `Spending` column with 3 visible groups, e.g. small spenders (~100–200), medium (~1,000–1,300) and large (~5,000–6,000), 3–4 values each. Apply `strategy='kmeans'` with 3 bins, print `bin_edges_`, and check whether each natural group ended up in its own bin.

*(Solution in the **Practice Solutions** section at the end.)*

In [14]:
# ✍️ Practice Problem 7 — write your solution here

### Solution 7

In [15]:
spend7 = pd.DataFrame({'Spending': [110, 140, 170, 200, 1000, 1100, 1250, 1300, 5000, 5400, 5900]})
kb7 = KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='kmeans')
spend7['Bin'] = kb7.fit_transform(spend7[['Spending']]).astype(int).ravel()
print('KBinsDiscretizer (kmeans) edges:', kb7.bin_edges_[0].round(0))
print(spend7)

# Check against a full K-Means with 10 different starts (as in Section 7)
from sklearn.cluster import KMeans
centres7 = np.sort(KMeans(n_clusters=3, n_init=10, random_state=0).fit(spend7[['Spending']]).cluster_centers_.ravel())
edges7 = (centres7[:-1] + centres7[1:]) / 2
print('\nFull K-Means centres:', centres7.round(0), '→ edges', edges7.round(0))
print('Bins with full K-Means:', np.digitize(spend7['Spending'], edges7).tolist())

KBinsDiscretizer (kmeans) edges: [ 110. 2929. 5550. 5900.]
    Spending  Bin
0        110    0
1        140    0
2        170    0
3        200    0
4       1000    0
5       1100    0
6       1250    0
7       1300    0
8       5000    1
9       5400    1
10      5900    2

Full K-Means centres: [ 155. 1162. 5433.] → edges [ 659. 3298.]
Bins with full K-Means: [0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2]


**Explanation:** You might expect one bin per natural group, but here `KBinsDiscretizer` merges the small and medium spenders into bin 0 and splits the large spenders into two bins. It is the same **local-optimum** pitfall as in Section 7: `KBinsDiscretizer` runs K-Means only once, starting from evenly spaced centres, and the big gap up to ~5,000 pulls the result the wrong way. A full `KMeans(n_init=10)` finds the three natural groups (≈ 155, ≈ 1,160, ≈ 5,430). Lesson: always **inspect** `bin_edges_` and the bins, and don't assume `kmeans` found the natural groups.

<a id="p-8"></a>
### ✍️ Practice Problem 8

Given `skewed = [1, 1, 2, 2, 3, 3, 4, 5, 6, 8, 12, 20, 45, 120, 400]`, which strategy would you **test first**, and why? Apply all three strategies with 4 bins, print the rows per bin for each, and write your conclusion in a comment.

*(Solution in the **Practice Solutions** section at the end.)*

In [16]:
# ✍️ Practice Problem 8 — write your solution here

### Solution 8

In [17]:
skewed8 = pd.DataFrame({'x': [1, 1, 2, 2, 3, 3, 4, 5, 6, 8, 12, 20, 45, 120, 400]})
for strat in ['uniform', 'quantile', 'kmeans']:
    kb8 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy=strat)
    counts8 = np.bincount(kb8.fit_transform(skewed8).astype(int).ravel(), minlength=4)
    print(f'{strat:<9} rows per bin = {counts8.tolist()} | edges = {kb8.bin_edges_[0].round(1)}')
# Test QUANTILE first: the data is strongly right-skewed, so uniform crowds almost everything into bin 0.

uniform   rows per bin = [13, 1, 0, 1] | edges = [  1.  100.8 200.5 300.2 400. ]
quantile  rows per bin = [2, 5, 4, 4] | edges = [  1.   2.   5.  20. 400.]
kmeans    rows per bin = [12, 1, 1, 1] | edges = [  1.   25.3  82.5 260.  400. ]


**Explanation:** For skewed data, quantile is the natural first choice. Uniform gives counts like [13, 1, 0, 1] (almost everything in bin 0, one empty bin). Kmeans gives [12, 1, 1, 1]: it spends three bins on the few extreme values. Quantile gives balanced groups [2, 5, 4, 4] (repeated values prevent a perfect split).

<a id="p-9"></a>
### ✍️ Practice Problem 9

Create an `Income` column with 12 values of your choice. Create **ordinal** bins with 4 quantile bins, add the bin IDs to the DataFrame, and add a text label column (`Low`, `Lower-mid`, `Upper-mid`, `High`).

*(Solution in the **Practice Solutions** section at the end.)*

In [18]:
# ✍️ Practice Problem 9 — write your solution here

### Solution 9

In [19]:
inc9 = pd.DataFrame({'Income': [15000, 22000, 28000, 33000, 39000, 45000, 52000, 60000, 72000, 88000, 110000, 180000]})
kb9 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile')
inc9['Income_Bin'] = kb9.fit_transform(inc9[['Income']]).astype(int).ravel()
inc9['Income_Label'] = inc9['Income_Bin'].map({0: 'Low', 1: 'Lower-mid', 2: 'Upper-mid', 3: 'High'})
inc9

,Income,Income_Bin,Income_Label
0,15000,0,Low
1,22000,0,Low
2,28000,0,Low
3,33000,1,Lower-mid
4,39000,1,Lower-mid
5,45000,1,Lower-mid
6,52000,2,Upper-mid
7,60000,2,Upper-mid
8,72000,2,Upper-mid
9,88000,3,High


**Explanation:** Three people per group. The bin IDs keep the order Low < … < High, but the income ranges behind them have very different widths.

<a id="p-10"></a>
### ✍️ Practice Problem 10

Create a DataFrame with a column `Hours_Studied` (10 values between 0 and 12). Use `encode='onehot'` with **3 bins**, convert the sparse result to a DataFrame with `get_feature_names_out`, and join it to the original data.

*(Solution in the **Practice Solutions** section at the end.)*

In [20]:
# ✍️ Practice Problem 10 — write your solution here

### Solution 10

In [21]:
hours10 = pd.DataFrame({'Hours_Studied': [0.5, 1, 2, 3, 4.5, 5, 6.5, 8, 10, 12]})
kb10 = KBinsDiscretizer(n_bins=3, encode='onehot', strategy='uniform')
sparse10 = kb10.fit_transform(hours10[['Hours_Studied']])
oh10 = pd.DataFrame(sparse10.toarray().astype(int), columns=kb10.get_feature_names_out(['Hours_Studied']))
pd.concat([hours10, oh10], axis=1)

,Hours_Studied,Hours_Studied_0.0,Hours_Studied_1.0,Hours_Studied_2.0
0,0.5,1,0,0
1,1.0,1,0,0
2,2.0,1,0,0
3,3.0,1,0,0
4,4.5,0,1,0
5,5.0,0,1,0
6,6.5,0,1,0
7,8.0,0,1,0
8,10.0,0,0,1
9,12.0,0,0,1


**Explanation:** `fit_transform` returns a sparse matrix; `.toarray()` makes it viewable, and `get_feature_names_out` gives readable column names.

<a id="p-11"></a>
### ✍️ Practice Problem 11

Using any numerical column (e.g. `Monthly_Bill` with 10 values), create a **dense** one-hot representation with **4 bins**, and put it into a DataFrame with proper column names.

*(Solution in the **Practice Solutions** section at the end.)*

In [22]:
# ✍️ Practice Problem 11 — write your solution here

### Solution 11

In [23]:
bill11 = pd.DataFrame({'Monthly_Bill': [900, 1200, 1500, 1800, 2200, 2600, 3100, 3700, 4400, 5200]})
kb11 = KBinsDiscretizer(n_bins=4, encode='onehot-dense', strategy='uniform')
dense11 = kb11.fit_transform(bill11[['Monthly_Bill']])            # already a normal NumPy array
print(type(dense11))
pd.concat([bill11, pd.DataFrame(dense11.astype(int), columns=kb11.get_feature_names_out(['Monthly_Bill']))], axis=1)

<class 'numpy.ndarray'>


,Monthly_Bill,Monthly_Bill_0.0,Monthly_Bill_1.0,Monthly_Bill_2.0,Monthly_Bill_3.0
0,900,1,0,0,0
1,1200,1,0,0,0
2,1500,1,0,0,0
3,1800,1,0,0,0
4,2200,0,1,0,0
5,2600,0,1,0,0
6,3100,0,0,1,0
7,3700,0,0,1,0
8,4400,0,0,0,1
9,5200,0,0,0,1


**Explanation:** No `.toarray()` needed — `onehot-dense` returns a regular array.

<a id="p-12"></a>
### ✍️ Practice Problem 12

Create a DataFrame with a `Price` column (10 values of your choice). Fit a `KBinsDiscretizer` with 3 quantile bins (**only `fit`**, no transform), then print `bin_edges_`, `n_bins_`, and each bin as a readable interval.

*(Solution in the **Practice Solutions** section at the end.)*

In [24]:
# ✍️ Practice Problem 12 — write your solution here

### Solution 12

In [25]:
price12 = pd.DataFrame({'Price': [120, 150, 175, 200, 260, 310, 400, 520, 700, 950]})
kb12 = KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='quantile')
kb12.fit(price12)                                     # learn only — nothing transformed yet
print('bin_edges_:', kb12.bin_edges_[0].round(1), '| n_bins_:', kb12.n_bins_)
e12 = kb12.bin_edges_[0]
for i in range(len(e12) - 1):
    print(f'bin {i}: {e12[i]:.1f} ≤ Price < {e12[i + 1]:.1f}')

bin_edges_: [120. 200. 400. 950.] | n_bins_: [3]
bin 0: 120.0 ≤ Price < 200.0
bin 1: 200.0 ≤ Price < 400.0
bin 2: 400.0 ≤ Price < 950.0


**Explanation:** `fit` only stores the edges; you need `transform` to get bin IDs.

<a id="p-13"></a>
### ✍️ Practice Problem 13

Split `pd.DataFrame({'Income': [20, 25, 30, 35, 40, 50, 60, 75, 90, 120, 150, 200]})` (in thousands) into the first 9 rows as train and the last 3 as test. Fit a 3-bin quantile discretizer on the **train** part only, then `transform` both parts and print the results.

*(Solution in the **Practice Solutions** section at the end.)*

In [26]:
# ✍️ Practice Problem 13 — write your solution here

### Solution 13

In [27]:
inc13 = pd.DataFrame({'Income': [20, 25, 30, 35, 40, 50, 60, 75, 90, 120, 150, 200]})
train13, test13 = inc13.iloc[:9], inc13.iloc[9:]
kb13 = KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='quantile').fit(train13)   # TRAIN only
print('Edges (from train):', kb13.bin_edges_[0])
print('Train bins:', kb13.transform(train13).astype(int).ravel())
print('Test bins :', kb13.transform(test13).astype(int).ravel(), '← all above the training max → last bin')

Edges (from train): [20. 35. 60. 90.]
Train bins: [0 0 0 1 1 1 2 2 2]
Test bins : [2 2 2] ← all above the training max → last bin


**Explanation:** The test incomes (120, 150, 200) are larger than anything in training, so they are clipped into the last bin — the edges are **not** re-learned.

<a id="p-14"></a>
### ✍️ Practice Problem 14

Create a DataFrame with two numerical columns (`Height_cm`, `Weight_kg`, 8 rows each). Use `fit_transform()` with 3 uniform ordinal bins on **both columns at once**, and put the result into a new DataFrame with columns `Height_Bin`, `Weight_Bin`.

*(Solution in the **Practice Solutions** section at the end.)*

In [28]:
# ✍️ Practice Problem 14 — write your solution here

### Solution 14

In [29]:
hw14 = pd.DataFrame({'Height_cm': [150, 158, 162, 167, 171, 176, 182, 190],
                     'Weight_kg': [48, 55, 60, 64, 70, 78, 85, 98]})
kb14 = KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='uniform')
binned14 = pd.DataFrame(kb14.fit_transform(hw14).astype(int), columns=['Height_Bin', 'Weight_Bin'])
pd.concat([hw14, binned14], axis=1)

,Height_cm,Weight_kg,Height_Bin,Weight_Bin
0,150,48,0,0
1,158,55,0,0
2,162,60,0,0
3,167,64,1,0
4,171,70,1,1
5,176,78,1,1
6,182,85,2,2
7,190,98,2,2


**Explanation:** One discretizer, two columns, separate edges for each.

<a id="p-15"></a>
### ✍️ Practice Problem 15

Find the data-leakage problem in this code and rewrite it correctly:
```python
kb = KBinsDiscretizer(n_bins=5, encode='ordinal', strategy='quantile')
X_all_binned = kb.fit_transform(customers[['Age', 'Income']])
X_train, X_test, y_train, y_test = train_test_split(X_all_binned, customers['Churn'], test_size=0.2, random_state=1)
model = LogisticRegression().fit(X_train, y_train)
print(model.score(X_test, y_test))
```
Create a small `customers` DataFrame (at least 40 rows, random data is fine) so your corrected code runs.

*(Solution in the **Practice Solutions** section at the end.)*

In [30]:
# ✍️ Practice Problem 15 — write your solution here

### Solution 15

In [31]:
from sklearn.linear_model import LogisticRegression
rng15s = np.random.default_rng(15)
customers15 = pd.DataFrame({'Age': rng15s.integers(18, 70, 120), 'Income': rng15s.integers(20000, 150000, 120)})
customers15['Churn'] = (rng15s.random(120) < 0.3).astype(int)

# PROBLEM: fit_transform on ALL rows before splitting → the test rows helped choose the bin edges (leakage).
# FIX: split first, fit on train, transform both.
X15 = customers15[['Age', 'Income']]; y15 = customers15['Churn']
X15_train, X15_test, y15_train, y15_test = train_test_split(X15, y15, test_size=0.2, random_state=1)
kb15 = KBinsDiscretizer(n_bins=5, encode='ordinal', strategy='quantile').fit(X15_train)
model15 = LogisticRegression().fit(kb15.transform(X15_train), y15_train)
print('Test accuracy:', round(model15.score(kb15.transform(X15_test), y15_test), 3))

# Even better — a Pipeline does it automatically:
pipe15 = Pipeline([('bins', KBinsDiscretizer(n_bins=5, encode='onehot', strategy='quantile')), ('model', LogisticRegression())])
print('Pipeline test accuracy:', round(pipe15.fit(X15_train, y15_train).score(X15_test, y15_test), 3))

Test accuracy: 0.792
Pipeline test accuracy: 0.792


**Explanation:** The fix is the order of operations: **split → fit on train → transform train and test**. (The target is random here, so accuracy itself is not meaningful — the workflow is the point.)

<a id="p-16"></a>
### ✍️ Practice Problem 16

Create a DataFrame with three numerical columns (`Age`, `Monthly_Income`, `Screen_Time_Hours`, 10 rows). Discretize all three with **4 uniform ordinal bins** using one `KBinsDiscretizer`, print the edges for each column, and show the binned DataFrame.

*(Solution in the **Practice Solutions** section at the end.)*

In [32]:
# ✍️ Practice Problem 16 — write your solution here

### Solution 16

In [33]:
df16 = pd.DataFrame({'Age': [21, 26, 30, 35, 40, 44, 49, 55, 60, 66],
                     'Monthly_Income': [30, 36, 45, 52, 60, 70, 78, 90, 110, 140],
                     'Screen_Time_Hours': [7.5, 6.8, 6.0, 5.2, 4.9, 4.1, 3.6, 3.0, 2.2, 1.5]})
kb16 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='uniform')
bins16 = pd.DataFrame(kb16.fit_transform(df16).astype(int), columns=[c + '_Bin' for c in df16.columns])
for col, e in zip(df16.columns, kb16.bin_edges_):
    print(f'{col:<18} {e.round(2)}')
pd.concat([df16, bins16], axis=1)

Age                [21.   32.25 43.5  54.75 66.  ]
Monthly_Income     [ 30.   57.5  85.  112.5 140. ]
Screen_Time_Hours  [1.5 3.  4.5 6.  7.5]


,Age,Monthly_Income,Screen_Time_Hours,Age_Bin,Monthly_Income_Bin,Screen_Time_Hours_Bin
0,21,30,7.5,0,0,3
1,26,36,6.8,0,0,3
2,30,45,6.0,0,0,3
3,35,52,5.2,1,0,2
4,40,60,4.9,1,1,2
5,44,70,4.1,2,1,1
6,49,78,3.6,2,1,1
7,55,90,3.0,3,2,1
8,60,110,2.2,3,2,0
9,66,140,1.5,3,3,0


**Explanation:** Each column has its own edges in its own units.

<a id="p-17"></a>
### ✍️ Practice Problem 17

Create a dataset with three numerical features (e.g. `Distance_km`, `Delivery_Time_min`, `Order_Value`, 12 rows). Discretize them with `n_bins=[2, 4, 3]` (uniform, ordinal) and confirm with `n_bins_` and `bin_edges_` that each column got the right number of bins.

*(Solution in the **Practice Solutions** section at the end.)*

In [34]:
# ✍️ Practice Problem 17 — write your solution here

### Solution 17

In [35]:
df17 = pd.DataFrame({'Distance_km': [1, 2, 3, 4, 6, 8, 10, 12, 15, 18, 22, 25],
                     'Delivery_Time_min': [10, 14, 18, 20, 25, 30, 34, 38, 45, 50, 58, 65],
                     'Order_Value': [300, 450, 500, 620, 700, 820, 950, 1100, 1300, 1600, 2000, 2500]})
kb17 = KBinsDiscretizer(n_bins=[2, 4, 3], encode='ordinal', strategy='uniform')
bins17 = kb17.fit_transform(df17).astype(int)
print('n_bins_:', kb17.n_bins_)
for col, e in zip(df17.columns, kb17.bin_edges_):
    print(f'{col:<18} {len(e) - 1} bins: {e.round(1)}')
pd.concat([df17, pd.DataFrame(bins17, columns=['Dist_Bin', 'Time_Bin', 'Value_Bin'])], axis=1)

n_bins_: [2 4 3]
Distance_km        2 bins: [ 1. 13. 25.]
Delivery_Time_min  4 bins: [10.  23.8 37.5 51.2 65. ]
Order_Value        3 bins: [ 300.  1033.3 1766.7 2500. ]


,Distance_km,Delivery_Time_min,Order_Value,Dist_Bin,Time_Bin,Value_Bin
0,1,10,300,0,0,0
1,2,14,450,0,0,0
2,3,18,500,0,0,0
3,4,20,620,0,0,0
4,6,25,700,0,1,0
5,8,30,820,0,1,0
6,10,34,950,0,1,0
7,12,38,1100,0,2,1
8,15,45,1300,1,2,1
9,18,50,1600,1,2,1


**Explanation:** The list is matched to the columns **in order**.

<a id="p-18"></a>
### ✍️ Practice Problem 18

Generate a skewed dataset: `rng = np.random.default_rng(1); spend = rng.exponential(scale=2000, size=200).round()`. Compare `uniform` and `quantile` with **4 bins**: print the edges and the counts per bin, and explain in a comment which is more useful and why.

*(Solution in the **Practice Solutions** section at the end.)*

In [36]:
# ✍️ Practice Problem 18 — write your solution here

### Solution 18

In [37]:
rng18 = np.random.default_rng(1)
spend18 = pd.DataFrame({'Spend': rng18.exponential(scale=2000, size=200).round()})
for strat in ['uniform', 'quantile']:
    kb18 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy=strat)
    c18 = np.bincount(kb18.fit_transform(spend18).astype(int).ravel(), minlength=4)
    print(f'{strat:<9} edges {kb18.bin_edges_[0].round(0).astype(int).tolist()} | counts {c18.tolist()}')
# Quantile is more useful: uniform puts most customers into the first bin because of a few very large spenders.

uniform   edges [14, 4222, 8430, 12638, 16846] | counts [176, 20, 3, 1]
quantile  edges [14, 632, 1506, 2733, 16846] | counts [50, 50, 50, 50]


**Explanation:** With exponential (right-skewed) data, uniform bins are dominated by the long tail; quantile gives 50 customers per bin.

<a id="p-19"></a>
### ✍️ Practice Problem 19

Take `Spending = [120, 150, 180, 200, 220, 260, 300, 350, 900, 2500, 8000, 15000]`. Create 4 equal-frequency groups **once with `pd.qcut()`** and **once with `KBinsDiscretizer(strategy='quantile')`**, show both results side by side, and say whether they agree.

*(Solution in the **Practice Solutions** section at the end.)*

In [38]:
# ✍️ Practice Problem 19 — write your solution here

### Solution 19

In [39]:
sp19 = pd.DataFrame({'Spending': [120, 150, 180, 200, 220, 260, 300, 350, 900, 2500, 8000, 15000]})
sp19['pd_qcut'] = pd.qcut(sp19['Spending'], q=4, labels=False)
sp19['kbins_quantile'] = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile').fit_transform(sp19[['Spending']]).astype(int).ravel()
print('Agree on every row?', (sp19['pd_qcut'] == sp19['kbins_quantile']).all())
sp19

Agree on every row? True


,Spending,pd_qcut,kbins_quantile
0,120,0,0
1,150,0,0
2,180,0,0
3,200,1,1
4,220,1,1
5,260,1,1
6,300,2,2
7,350,2,2
8,900,2,2
9,2500,3,3


**Explanation:** Both create equal-frequency groups. They usually agree; tiny differences can appear at exact edges because of different percentile methods/interval closing.

<a id="p-20"></a>
### ✍️ Practice Problem 20

For each task, write in a comment whether it needs **scaling**, **discretization**, **both**, or **neither**, and why:
1. KNN model using Age (18–80) and Income (20,000–200,000).
2. A marketing report that needs "Low / Medium / High spender" groups.
3. A Random Forest predicting churn from monthly charges.
4. Logistic regression where the effect of age on risk is U-shaped (young and old are riskier).
5. A neural network with features on very different scales.

*(Solution in the **Practice Solutions** section at the end.)*

In [40]:
# ✍️ Practice Problem 20 — write your solution here

### Solution 20

In [41]:
answers20 = {
    '1. KNN with Age & Income': 'SCALING — KNN uses distances; Income would dominate Age without scaling.',
    '2. Low/Medium/High spender report': 'DISCRETIZATION — the goal is interpretable groups.',
    '3. Random Forest on monthly charges': 'NEITHER is needed — trees are scale-invariant and find their own split points.',
    '4. Logistic regression, U-shaped age effect': 'DISCRETIZATION (one-hot bins) — lets a linear model capture high-low-high risk (plus scaling of other features).',
    '5. Neural network, features on different scales': 'SCALING — gradient-based training works best with similar scales.',
}
for k, v in answers20.items():
    print(f'{k:<45} → {v}')

1. KNN with Age & Income                      → SCALING — KNN uses distances; Income would dominate Age without scaling.
2. Low/Medium/High spender report             → DISCRETIZATION — the goal is interpretable groups.
3. Random Forest on monthly charges           → NEITHER is needed — trees are scale-invariant and find their own split points.
4. Logistic regression, U-shaped age effect   → DISCRETIZATION (one-hot bins) — lets a linear model capture high-low-high risk (plus scaling of other features).
5. Neural network, features on different scales → SCALING — gradient-based training works best with similar scales.


**Explanation:** Scaling changes the scale (keeps information); discretization creates groups (loses some information on purpose).

<a id="p-21"></a>
### ✍️ Practice Problem 21

Create an `Age` column (12 values between 18 and 80). Discretize it into 4 quantile bins and then convert the bins into **one-hot encoded features** — once using `encode='onehot-dense'` directly, and once using `encode='ordinal'` followed by `OneHotEncoder`. Show both results as DataFrames.

*(Solution in the **Practice Solutions** section at the end.)*

In [42]:
# ✍️ Practice Problem 21 — write your solution here

### Solution 21

In [43]:
age21 = pd.DataFrame({'Age': [18, 22, 26, 31, 35, 40, 46, 52, 57, 63, 70, 78]})
# Way 1 — directly
kb21 = KBinsDiscretizer(n_bins=4, encode='onehot-dense', strategy='quantile')
way1 = pd.DataFrame(kb21.fit_transform(age21).astype(int), columns=kb21.get_feature_names_out(['Age']))
# Way 2 — ordinal bins, then OneHotEncoder
ord21 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile').fit_transform(age21)
enc21 = OneHotEncoder(sparse_output=False)
way2 = pd.DataFrame(enc21.fit_transform(ord21).astype(int), columns=[f'AgeBin_{int(c)}' for c in enc21.categories_[0]])
print('Same one-hot values?', np.array_equal(way1.to_numpy(), way2.to_numpy()))
pd.concat([age21, way1, way2], axis=1)

Same one-hot values? True


,Age,Age_0.0,Age_1.0,Age_2.0,Age_3.0,AgeBin_0,AgeBin_1,AgeBin_2,AgeBin_3
0,18,1,0,0,0,1,0,0,0
1,22,1,0,0,0,1,0,0,0
2,26,1,0,0,0,1,0,0,0
3,31,0,1,0,0,0,1,0,0
4,35,0,1,0,0,0,1,0,0
5,40,0,1,0,0,0,1,0,0
6,46,0,0,1,0,0,0,1,0
7,52,0,0,1,0,0,0,1,0
8,57,0,0,1,0,0,0,1,0
9,63,0,0,0,1,0,0,0,1


**Explanation:** Both routes give identical one-hot columns; the two-step route lets you use `OneHotEncoder` options such as `drop='first'`.

<a id="p-22"></a>
### ✍️ Practice Problem 22

In a comment, explain in your own words **why** a decision tree can find useful bin boundaries based on the target variable, and why this is different from `KBinsDiscretizer`. Then, using the `credit` DataFrame above, fit a tree with `max_leaf_nodes=4` and print its thresholds.

*(Solution in the **Practice Solutions** section at the end.)*

In [44]:
# ✍️ Practice Problem 22 — write your solution here

### Solution 22

In [45]:
# Why it works (in words):
# A decision tree tries many possible thresholds on Age and keeps the one that makes the Default groups PUREST
# (largest information gain). So its thresholds sit where the default rate CHANGES.
# KBinsDiscretizer never looks at the target — it only uses the spread of Age (range, percentiles, clusters).
tree22 = DecisionTreeClassifier(max_leaf_nodes=4, min_samples_leaf=30, random_state=0).fit(credit[['Age']], credit['Default'])
th22 = sorted(t for t in tree22.tree_.threshold if t != -2)
print('Tree thresholds:', np.round(th22, 1))
print('Default rate per tree bin:', credit.groupby(np.digitize(credit['Age'], th22))['Default'].mean().round(3).to_dict())

Tree thresholds: [23.5 60.5 65.5]
Default rate per tree bin: {0: 0.379, 1: 0.111, 2: 0.365, 3: 0.226}


**Explanation:** With 4 leaves the tree finds 3 thresholds; they cluster around the ages where risk changes (≈ 24 and ≈ 60).

<a id="p-23"></a>
### ✍️ Practice Problem 23

Twelve customers, target = `Bought` (1/0):
```python
d23 = pd.DataFrame({'Income': [20, 25, 28, 30, 35, 40, 55, 60, 70, 80, 90, 100],
                    'Bought': [0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1]})
```
Compare the splits **Income < 32**, **Income < 50** and **Income < 85**: count the classes on each side, compute the Gini information gain for each, and say which split separates the classes best.

*(Solution in the **Practice Solutions** section at the end.)*

In [46]:
# ✍️ Practice Problem 23 — write your solution here

### Solution 23

In [47]:
d23 = pd.DataFrame({'Income': [20, 25, 28, 30, 35, 40, 55, 60, 70, 80, 90, 100],
                    'Bought': [0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1]})
parent23 = d23['Bought'].value_counts().reindex([0, 1]).tolist()
for thr in [32, 50, 85]:
    left = d23.loc[d23['Income'] < thr, 'Bought'].value_counts().reindex([0, 1], fill_value=0).tolist()
    right = d23.loc[d23['Income'] >= thr, 'Bought'].value_counts().reindex([0, 1], fill_value=0).tolist()
    print(f'Income < {thr}: left (no, yes) = {left}, right = {right} → Gini gain = {information_gain(parent23, [left, right]):.3f}')

Income < 32: left (no, yes) = [3, 1], right = [3, 5] → Gini gain = 0.062
Income < 50: left (no, yes) = [5, 1], right = [1, 5] → Gini gain = 0.222
Income < 85: left (no, yes) = [5, 5], right = [1, 1] → Gini gain = 0.000


**Explanation:** `Income < 50` gives the largest gain: the left side is 5 × no / 1 × yes and the right side is 1 × no / 5 × yes — much purer than the other splits.

<a id="p-24"></a>
### ✍️ Practice Problem 24

Create a `KBinsDiscretizer` with `dtype=np.float32` (4 quantile bins, ordinal) and apply it to a column of 500,000 random values. Compare the memory (`.nbytes`) with a `float64` version and explain in a comment why memory usage matters for large datasets. Also try `dtype='float32'` and note what happens.

*(Solution in the **Practice Solutions** section at the end.)*

In [48]:
# ✍️ Practice Problem 24 — write your solution here

### Solution 24

In [49]:
vals24 = pd.DataFrame({'x': np.random.default_rng(24).normal(size=500_000)})
out64_24 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile', dtype=np.float64).fit_transform(vals24)
out32_24 = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile', dtype=np.float32).fit_transform(vals24)
print(f'float64: {out64_24.nbytes / 1e6:.1f} MB | float32: {out32_24.nbytes / 1e6:.1f} MB')
try:
    KBinsDiscretizer(dtype='float32').fit(vals24)
except Exception as e:
    print('dtype="float32" →', type(e).__name__, ': the string is not accepted; pass np.float32')
# Memory matters because large datasets (millions of rows × many columns) may not fit in RAM;
# float32 halves memory and is precise enough for bin IDs 0, 1, 2, 3.

float64: 4.0 MB | float32: 2.0 MB
dtype="float32" → InvalidParameterError : the string is not accepted; pass np.float32


**Explanation:** Bin IDs are small whole numbers, so `float32` loses nothing and uses half the memory.

<a id="p-25"></a>
### ✍️ Practice Problem 25

Create a dataset with a numeric column `Age` containing at least 3 `NaN` values (12 rows). Build a `Pipeline` that (1) imputes with the median and (2) discretizes into 3 uniform ordinal bins. Show the imputed value and the final bins.

*(Solution in the **Practice Solutions** section at the end.)*

In [50]:
# ✍️ Practice Problem 25 — write your solution here

### Solution 25

In [51]:
age25 = pd.DataFrame({'Age': [22, np.nan, 31, 36, np.nan, 44, 51, 58, np.nan, 63, 70, 27]})
pipe25 = Pipeline([('impute', SimpleImputer(strategy='median')),
                   ('bins', KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='uniform'))])
out25 = pipe25.fit_transform(age25)
print('Imputed median:', pipe25.named_steps['impute'].statistics_[0])
age25.assign(Age_Bin=out25.astype(int).ravel(), Was_Missing=age25['Age'].isna().astype(int))

Imputed median: 44.0


,Age,Age_Bin,Was_Missing
0,22.0,0,0
1,NaN,1,1
2,31.0,0,0
3,36.0,0,0
4,NaN,1,1
5,44.0,1,0
6,51.0,1,0
7,58.0,2,0
8,NaN,1,1
9,63.0,2,0


**Explanation:** Impute first, then bin; keep a `Was_Missing` flag if missingness might be informative.

<a id="p-26"></a>
### ✍️ Practice Problem 26

Start with `scores = [45, 50, 52, 55, 58, 60, 62, 65, 68, 70]` and bin them with 3 **uniform** bins. Then add one extreme value `500` and bin again. Print the edges and bins before and after, and explain in a comment how one extreme value changed the result.

*(Solution in the **Practice Solutions** section at the end.)*

In [52]:
# ✍️ Practice Problem 26 — write your solution here

### Solution 26

In [53]:
scores26 = [45, 50, 52, 55, 58, 60, 62, 65, 68, 70]
for label, data26 in [('before', scores26), ('after adding 500', scores26 + [500])]:
    kb26 = KBinsDiscretizer(n_bins=3, encode='ordinal', strategy='uniform')
    b26 = kb26.fit_transform(pd.DataFrame({'s': data26})).astype(int).ravel()
    print(f'{label:<17} edges {kb26.bin_edges_[0].round(1)} | bins {b26.tolist()}')
# One extreme value stretched the range from 45–70 to 45–500, so all normal scores collapsed into bin 0
# and the middle bin became empty — uniform binning is very sensitive to outliers.

before            edges [45.  53.3 61.7 70. ] | bins [0, 0, 0, 1, 1, 1, 2, 2, 2, 2]
after adding 500  edges [ 45.  196.7 348.3 500. ] | bins [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]


**Explanation:** Before the outlier the ten scores are spread over three bins; afterwards all ten land in bin 0.

<a id="p-27"></a>
### ✍️ Practice Problem 27

Create your own pipeline with `KBinsDiscretizer` and a classification model (e.g. `LogisticRegression` or `DecisionTreeClassifier`) on the breast-cancer data (`from sklearn.datasets import load_breast_cancer`; use 3–5 numeric columns). Report 5-fold CV accuracy with and without the discretizer.

*(Solution in the **Practice Solutions** section at the end.)*

In [54]:
# ✍️ Practice Problem 27 — write your solution here

### Solution 27

In [55]:
from sklearn.datasets import load_breast_cancer
bc27 = load_breast_cancer(as_frame=True)
X27 = bc27.data[['mean radius', 'mean texture', 'mean concavity', 'mean smoothness']]; y27 = bc27.target
cv27 = StratifiedKFold(5, shuffle=True, random_state=0)
pipes27 = {
    'LogReg (scaled, continuous)': Pipeline([('scale', StandardScaler()), ('model', LogisticRegression(max_iter=1000))]),
    'KBins (5 quantile, one-hot) + LogReg': Pipeline([('bins', KBinsDiscretizer(n_bins=5, encode='onehot', strategy='quantile')),
                                                      ('model', LogisticRegression(max_iter=1000))]),
}
for name, p in pipes27.items():
    print(f'{name:<38} CV accuracy = {cross_val_score(p, X27, y27, cv=cv27).mean():.3f}')

LogReg (scaled, continuous)            CV accuracy = 0.937


KBins (5 quantile, one-hot) + LogReg   CV accuracy = 0.926


**Explanation:** Both pipelines are leakage-safe. Here the continuous version is typically a little better: these features have smooth, roughly monotonic relationships with the target, so binning mostly loses information.

<a id="p-28"></a>
### ✍️ Practice Problem 28

Build a similar end-to-end workflow with **your own** small dataset (at least 200 rows, 3–4 numeric features, a binary target). Use a `ColumnTransformer` with at least one `KBinsDiscretizer`, a classification model, a train/test split, and compare with a no-binning baseline using cross-validation.

*(Solution in the **Practice Solutions** section at the end.)*

In [56]:
# ✍️ Practice Problem 28 — write your solution here

### Solution 28

In [57]:
rng28s = np.random.default_rng(99)
own = pd.DataFrame({'Tenure_Months': rng28s.integers(1, 72, 250), 'Monthly_Charge': rng28s.uniform(20, 120, 250).round(1),
                    'Support_Calls': rng28s.poisson(2, 250)})
own['Churn'] = (rng28s.random(250) < 1 / (1 + np.exp(-(-0.5 - 0.04 * own['Tenure_Months'] + 0.4 * own['Support_Calls'])))).astype(int)
Xo, yo = own.drop(columns='Churn'), own['Churn']
Xo_tr, Xo_te, yo_tr, yo_te = train_test_split(Xo, yo, test_size=0.25, stratify=yo, random_state=0)

binned_pipe = Pipeline([('prep', ColumnTransformer([
                            ('tenure_bins', KBinsDiscretizer(n_bins=4, encode='onehot', strategy='quantile'), ['Tenure_Months']),
                            ('scale', StandardScaler(), ['Monthly_Charge', 'Support_Calls'])])),
                        ('model', LogisticRegression(max_iter=1000))])
base_pipe = Pipeline([('scale', StandardScaler()), ('model', LogisticRegression(max_iter=1000))])
for name, p in [('Baseline (no binning)', base_pipe), ('Tenure binned', binned_pipe)]:
    cv_s = cross_val_score(p, Xo_tr, yo_tr, cv=StratifiedKFold(5, shuffle=True, random_state=0), scoring='roc_auc').mean()
    test_s = roc_auc_score(yo_te, p.fit(Xo_tr, yo_tr).predict_proba(Xo_te)[:, 1])
    print(f'{name:<22} CV ROC-AUC = {cv_s:.3f} | test ROC-AUC = {test_s:.3f}')

Baseline (no binning)  CV ROC-AUC = 0.727 | test ROC-AUC = 0.803


Tenure binned          CV ROC-AUC = 0.729 | test ROC-AUC = 0.751


**Explanation:** The same workflow as Section 28. Keep the discretization only if cross-validation shows it helps, or if you need interpretable groups.

<a id="p-29"></a>
---
# Section 31 — Interview Questions

Try answering out loud first. **Answers are in the "Interview Answers" section at the end.**

1. What is discretization?
2. What is binning?
3. What is `KBinsDiscretizer`?
4. What does `n_bins` mean?
5. What does `strategy='uniform'` do?
6. What does `strategy='quantile'` do?
7. What does `strategy='kmeans'` do?
8. What is the difference between `pd.cut()` and `pd.qcut()`?
9. What is the difference between `pd.cut()` and `KBinsDiscretizer`?
10. What is the difference between `fit()` and `transform()`?
11. What is `bin_edges_`?
12. What is data leakage?
13. Why should we fit the discretizer on training data only?
14. What is ordinal encoding of bins, and what is its risk?
15. What is one-hot encoding of bins, and what is the difference between `onehot` and `onehot-dense`?
16. What are the disadvantages of discretization?
17. When should discretization be avoided?
18. What happens to values outside the training range during `transform()`?
19. Why might you get fewer bins than you asked for?
20. What is supervised discretization, and how can a decision tree be used for it?

---
# Interview Answers

| # | Question | Answer |
|---|---|---|
| 1 | What is discretization? | Converting a continuous numerical variable into a finite number of intervals (bins/groups). |
| 2 | What is binning? | The practical process of choosing interval boundaries and assigning each value to the interval it falls in; used as a synonym for discretization. |
| 3 | What is `KBinsDiscretizer`? | A scikit-learn transformer that learns bin edges for each numeric feature (uniform, quantile or kmeans) and outputs bin IDs or one-hot encodings; works with `fit`/`transform` and pipelines. |
| 4 | What does `n_bins` mean? | The number of bins per feature (an int, or a list with one value per feature). Must be ≥ 2; fewer bins may be used if edges are duplicates (`n_bins_`). |
| 5 | What is `uniform`? | Equal-**width** bins: (max − min) / n_bins. Simple; sensitive to skew and outliers. |
| 6 | What is `quantile`? | Equal-**frequency** bins: edges at percentiles so each bin has about the same number of observations. Good for skewed data. |
| 7 | What is `kmeans`? | Runs 1-D K-Means; edges are midpoints between neighbouring cluster centres, so values close to each other share a bin. Good for natural groups; results depend on data (single run can be a local optimum). |
| 8 | `pd.cut()` vs `pd.qcut()`? | `cut` = equal-width bins or custom edges; `qcut` = equal-frequency (quantile) bins. |
| 9 | `pd.cut()` vs `KBinsDiscretizer`? | `pd.cut` is a Pandas one-off function (supports custom edges and labels); `KBinsDiscretizer` learns edges on training data, reuses them on new data, offers kmeans and one-hot output, and fits into ML pipelines. |
| 10 | `fit()` vs `transform()`? | `fit` learns the bin edges from data; `transform` applies stored edges to (any) data; `fit_transform` does both on the same data (training only). |
| 11 | What is `bin_edges_`? | The learned boundaries: an array per feature with `n_bins + 1` values; bin *i* is `[edge_i, edge_{i+1})`, the last bin includes its right edge. |
| 12 | What is data leakage? | When information unavailable at prediction time — e.g. the test set — influences training or preprocessing, producing over-optimistic evaluation. |
| 13 | Why fit on training data only? | So the edges reflect only "seen" data; the test set then honestly simulates new data. Fitting on all data lets test values shape the boundaries. |
| 14 | Ordinal encoding & its risk? | Each value becomes a single bin ID (0, 1, 2…). Risk: models like linear regression treat the IDs as numbers with equal spacing, which may not be true. |
| 15 | One-hot encoding; `onehot` vs `onehot-dense`? | One 0/1 column per bin. `onehot` returns a memory-efficient sparse matrix; `onehot-dense` returns a regular NumPy array. |
| 16 | Disadvantages of discretization? | Information loss, boundary effects, arbitrary edges, more columns (one-hot), possible overfitting with many bins, often worse for tree models. |
| 17 | When to avoid discretization? | When exact values matter, with tree-based models, with small data, when boundaries would be arbitrary, or when scaling/transformation solves the real problem. |
| 18 | Values outside the training range? | They are clipped: below the first edge → bin 0; above the last edge → last bin. |
| 19 | Why fewer bins than requested? | Duplicate edges (many repeated values, or constant data) are removed with a warning; check `n_bins_`. |
| 20 | Supervised discretization? | Using the target to choose edges, e.g. fitting a small decision tree on one feature and taking its split thresholds (chosen by information gain/Gini). Must be fitted on training data only and validated to avoid overfitting. |

<a id="p-30"></a>
---
# Section 32 — Final Practice Assignment

**Scenario:** a company has customer data with these columns:
```text
CustomerID, Age, Income, SpendingScore, YearsAsCustomer, Purchased
```
Use this code to create the data (300 customers):
```python
rng_f = np.random.default_rng(2025)
n = 300
final_df = pd.DataFrame({
    'CustomerID': [f'C{1000 + i}' for i in range(n)],
    'Age': rng_f.integers(18, 70, n),
    'Income': np.round(rng_f.lognormal(10.8, 0.6, n), -2),
    'SpendingScore': rng_f.integers(1, 101, n),
    'YearsAsCustomer': np.round(rng_f.exponential(4, n), 1),
})
logit_f = (-1.2 + 0.03 * (final_df['SpendingScore'] - 50) + 0.8 * (final_df['YearsAsCustomer'] > 3)
           + 0.6 * (final_df['Income'] > 60000) - 0.5 * (final_df['Age'] > 55))
final_df['Purchased'] = (rng_f.random(n) < 1 / (1 + np.exp(-logit_f))).astype(int)
```

**Tasks**
1. Create the DataFrame.
2. Inspect the data (shape, types, missing values, summary, target rate).
3. Identify the continuous variables (and which column must **not** be used as a feature).
4. Split X and y.
5. Train/test split (25% test, stratified, `random_state=42`).
6. Choose a discretization strategy for each continuous feature.
7. Explain **why** you selected it (look at skewness and distributions).
8. Create **4 bins** per feature.
9. Use **ordinal** encoding.
10. Inspect the learned bin boundaries (`bin_edges_`).
11. Train a simple ML model (e.g. `DecisionTreeClassifier(max_depth=4)` or `LogisticRegression`).
12. Evaluate the model on the test set (accuracy, ROC-AUC, confusion matrix).
13. Try **another** discretization strategy.
14. Compare the results (also compare with a model using the continuous features).
15. Explain whether discretization helped.

Use the empty cell below. **The complete solution is at the very end.**

In [58]:
# ✍️ Final Practice Assignment — your solution here

---
# Final Practice Assignment Solution

### 1. Create the DataFrame · 2. Inspect the data

In [59]:
rng_f = np.random.default_rng(2025)
n = 300
final_df = pd.DataFrame({
    'CustomerID': [f'C{1000 + i}' for i in range(n)],
    'Age': rng_f.integers(18, 70, n),
    'Income': np.round(rng_f.lognormal(10.8, 0.6, n), -2),
    'SpendingScore': rng_f.integers(1, 101, n),
    'YearsAsCustomer': np.round(rng_f.exponential(4, n), 1),
})
logit_f = (-1.2 + 0.03 * (final_df['SpendingScore'] - 50) + 0.8 * (final_df['YearsAsCustomer'] > 3)
           + 0.6 * (final_df['Income'] > 60000) - 0.5 * (final_df['Age'] > 55))
final_df['Purchased'] = (rng_f.random(n) < 1 / (1 + np.exp(-logit_f))).astype(int)

print('Shape:', final_df.shape)
print(final_df.dtypes)
print('\nMissing values:', final_df.isna().sum().sum())
print('Purchase rate :', final_df['Purchased'].mean().round(3))
print('\nSkewness:'); print(final_df[['Age', 'Income', 'SpendingScore', 'YearsAsCustomer']].skew().round(2))
final_df.describe().round(1)

Shape: (300, 6)
CustomerID             str
Age                  int64
Income             float64
SpendingScore        int64
YearsAsCustomer    float64
Purchased            int64
dtype: object

Missing values: 0
Purchase rate : 0.337

Skewness:
Age                0.10
Income             3.05
SpendingScore      0.03
YearsAsCustomer    2.51
dtype: float64


,Age,Income,SpendingScore,YearsAsCustomer,Purchased
count,300.0,300.0,300.0,300.0,300.0
mean,43.6,58739.3,49.5,3.9,0.3
std,15.0,41515.0,28.7,4.1,0.5
min,18.0,10100.0,1.0,0.0,0.0
25%,31.8,32675.0,24.0,1.2,0.0
50%,43.0,49600.0,50.0,2.6,0.0
75%,56.0,71600.0,74.2,5.3,1.0
max,69.0,341000.0,100.0,33.3,1.0


### 3. Identify continuous variables
- Continuous features: `Age`, `Income`, `SpendingScore`, `YearsAsCustomer`.
- `CustomerID` is an **identifier** → never a feature.
- `Purchased` is the **target**.

### 4. Split X and y · 5. Train/test split

In [60]:
features_f = ['Age', 'Income', 'SpendingScore', 'YearsAsCustomer']
X_f = final_df[features_f]
y_f = final_df['Purchased']
Xf_train, Xf_test, yf_train, yf_test = train_test_split(X_f, y_f, test_size=0.25, stratify=y_f, random_state=42)
print('Train:', Xf_train.shape, '| Test:', Xf_test.shape)

Train: (225, 4) | Test: (75, 4)


### 6–7. Choose a discretization strategy (and why)
- `Income` (skew ≈ 3.0) and `YearsAsCustomer` (skew ≈ 2.5) are clearly **right-skewed** → uniform bins would crowd most customers into bin 0.
- `Age` and `SpendingScore` are roughly evenly spread.
- **Choice: `quantile`** for all four features. It gives balanced bins for the skewed features and is still sensible for the even ones. It is also a good default when we want every bin to have enough customers.

### 8–10. Create 4 bins, ordinal encoding, inspect the boundaries

In [61]:
kb_final = KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile')
kb_final.fit(Xf_train)                                   # learn edges on TRAINING data only
for col, e in zip(features_f, kb_final.bin_edges_):
    print(f'{col:<16} edges: {e.round(1)}')

Xf_train_b = kb_final.transform(Xf_train)
Xf_test_b = kb_final.transform(Xf_test)
print('\nCustomers per bin (train):')
print(pd.DataFrame(Xf_train_b.astype(int), columns=features_f).apply(lambda c: c.value_counts().sort_index()))

Age              edges: [18. 32. 43. 57. 69.]
Income           edges: [ 11800.  35300.  50200.  70400. 307300.]
SpendingScore    edges: [  1.  24.  51.  74. 100.]
YearsAsCustomer  edges: [ 0.   1.2  2.7  5.4 20.1]

Customers per bin (train):
   Age  Income  SpendingScore  YearsAsCustomer
0   56      56             55               53
1   51      56             56               58
2   58      56             57               56
3   60      57             57               58


### 11. Train a simple ML model · 12. Evaluate

In [62]:
tree_f = DecisionTreeClassifier(max_depth=4, random_state=0).fit(Xf_train_b, yf_train)
pred_f = tree_f.predict(Xf_test_b)
proba_f = tree_f.predict_proba(Xf_test_b)[:, 1]
print('Quantile bins + decision tree')
print('  accuracy:', round(accuracy_score(yf_test, pred_f), 3), '| ROC-AUC:', round(roc_auc_score(yf_test, proba_f), 3))
print('  confusion matrix:\n', confusion_matrix(yf_test, pred_f))

Quantile bins + decision tree
  accuracy: 0.693 | ROC-AUC: 0.743
  confusion matrix:
 [[37 13]
 [10 15]]


### 13–14. Try another strategy and compare (with pipelines + cross-validation for a fair comparison)

In [63]:
cv_f = StratifiedKFold(5, shuffle=True, random_state=0)
candidates_f = {
    'Continuous (no bins) + tree': DecisionTreeClassifier(max_depth=4, random_state=0),
    'Quantile bins + tree': Pipeline([('bins', KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='quantile')),
                                      ('tree', DecisionTreeClassifier(max_depth=4, random_state=0))]),
    'Uniform bins + tree': Pipeline([('bins', KBinsDiscretizer(n_bins=4, encode='ordinal', strategy='uniform')),
                                     ('tree', DecisionTreeClassifier(max_depth=4, random_state=0))]),
    'Continuous + LogReg (scaled)': Pipeline([('scale', StandardScaler()), ('lr', LogisticRegression(max_iter=1000))]),
    'Quantile one-hot + LogReg': Pipeline([('bins', KBinsDiscretizer(n_bins=4, encode='onehot', strategy='quantile')),
                                           ('lr', LogisticRegression(max_iter=1000))]),
}
rows_f = []
for name, m in candidates_f.items():
    cv_auc = cross_val_score(m, Xf_train, yf_train, cv=cv_f, scoring='roc_auc')
    m.fit(Xf_train, yf_train)
    rows_f.append({'Model': name, 'CV ROC-AUC': cv_auc.mean(), 'CV std': cv_auc.std(),
                   'Test ROC-AUC': roc_auc_score(yf_test, m.predict_proba(Xf_test)[:, 1]),
                   'Test accuracy': accuracy_score(yf_test, m.predict(Xf_test))})
pd.DataFrame(rows_f).set_index('Model').sort_values('CV ROC-AUC', ascending=False).round(3)

,CV ROC-AUC,CV std,Test ROC-AUC,Test accuracy
Model,,,,
Continuous + LogReg (scaled),0.720,0.055,0.748,0.707
Quantile one-hot + LogReg,0.692,0.066,0.729,0.733
Quantile bins + tree,0.673,0.056,0.743,0.693
Uniform bins + tree,0.651,0.064,0.737,0.747
Continuous (no bins) + tree,0.617,0.093,0.621,0.640


### 15. Did discretization help? — interpreting the results

(The numbers are reproducible because of the fixed random seeds.)

| Model | CV ROC-AUC (± std) | What it tells us |
|---|---|---|
| Continuous + LogReg (scaled) | ≈ 0.72 ± 0.06 | **Best CV score** |
| Quantile one-hot + LogReg | ≈ 0.69 ± 0.07 | Slightly lower, within noise |
| Quantile bins + tree | ≈ 0.67 ± 0.06 | Clearly better than the tree on raw values |
| Uniform bins + tree | ≈ 0.65 ± 0.06 | A bit behind quantile (skewed features) |
| Continuous + tree | ≈ 0.62 ± 0.09 | Worst and most unstable |

**Conclusions**
- **For the decision tree, discretization helped.** With only 225 training rows, a tree on raw values picks very specific, noisy thresholds that don't generalise. Four bins per feature act as **regularization**: the tree can only split at a few sensible places. Quantile bins did a little better than uniform bins, which matches the skewed `Income` and `YearsAsCustomer` (within noise, though).
- **For logistic regression, discretization did not help.** The scaled continuous features scored slightly higher. The one-hot bins can express the step effects, but they lose detail inside each bin, and here that roughly cancels the benefit.
- **Best overall:** logistic regression on continuous features, with the binned models close behind. The differences between the top models are about one CV standard deviation, so they are not decisive.
- **Test-set numbers** (75 rows) bounce around a lot; use the **CV** columns for decisions.

**Final answer:** *"Quantile discretization was chosen because two features are strongly right-skewed. It clearly improved the decision tree, mainly by regularizing it, but it did not beat logistic regression on scaled continuous features. The bins remain useful for explaining the model, e.g. 'customers in the top income quartile buy more'."*